### PR TITLE
[c10d] Make barrier as a custom op

### PR DESCRIPTION
**Pull request by alanwaketan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #79779
* __->__ #79777
* #79691

Summary:
This patch makes barrier as a custom op such that it's dispatcher
passable. It's one part of the effort to route comm ops to the dispatcher
such that tracing mechanisms that relies on the dispatcher can trace them,
e.g., LazyTensor and AOTAutograd.

Test Plan:
python test/distributed/test_c10d_nccl.py -k test_nccl_barrier
...and other existing distributed tests.

### CODE REVIEW

File: torch/csrc/distributed/c10d/Ops.cpp [hunk 1]
```diff
@@ -142,6 +150,9 @@ TORCH_LIBRARY(c10d, m) {
   m.def(
       "alltoall_",
       dispatch(c10::DispatchKey::CompositeExplicitAutograd, alltoall_));
+  m.def(
+      "barrier",
```
Comment (mrshenli) [inline torch/csrc/distributed/c10d/Ops.cpp:154]: missing a trailing underscore?

File: torch/csrc/distributed/c10d/Ops.cpp [hunk 2]
```diff
@@ -110,6 +110,14 @@ c10::intrusive_ptr<ProcessGroup::Work> alltoall_(
       AllToAllOptions{std::chrono::milliseconds(timeout)});
 }
 
+c10::intrusive_ptr<ProcessGroup::Work> barrier(
```
[the remark this answers is not in the document]
Comment (alanwaketan) [reply torch/csrc/distributed/c10d/Ops.cpp:113]: barrier is not a in-place op. Therefore, it's not necessary to add a trailing _.

File: torch/csrc/distributed/c10d/Ops.cpp (hunk 1 above)
Comment (alanwaketan) [reply torch/csrc/distributed/c10d/Ops.cpp:154]: Ditto.

File: torch/csrc/distributed/c10d/Ops.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (mrshenli) [reply torch/csrc/distributed/c10d/Ops.cpp:113]: Oh I see, trailing score is for inplace ops as in Python APIs. Thought it was a dispatcher convention.

**Review by mrshenli — APPROVED** [review]

**Review by wanchaol — APPROVED** [review]

**Pull request closed without merging**